<launch>
  <arg name="INPUT_IMAGE" default="/multisense/left/image_rect_color" />
  <arg name="INPUT_CAMERA_INFO" default="/multisense/left/camera_info" />
  <arg name="INPUT_POINT_CLOUD" default="/multisense/resize_1_4/points" />
  <arg name="CAMERA_FRAME" default="multisense/left_camera_optical_frame" />
  <arg name="DEBUG" default="true" />
  <arg name="LAUNCH_DYNAMIC_TF_PUBLISHER" default="true" />
  <group ns="drive">
    <node pkg="nodelet" type="nodelet" name="manager" args="manager" />
    <node pkg="image_view2" type="image_view2"  name="image_view2" if="$(arg DEBUG)">
      <remap from="image" to="$(arg INPUT_IMAGE)" />
    </node>
    <node pkg="image_view" type="image_view" name="mask_image_view" if="$(arg DEBUG)">
      <remap from="image" to="rect_to_mask_image/output" />
    </node>
    <node pkg="nodelet" type="nodelet" name="rect_to_mask_image"
          args="load jsk_perception/RectToMaskImage manager">
      <remap from="~input" to="$(arg INPUT_IMAGE)/screenrectangle" />
      <remap from="~input/camera_info" to="$(arg INPUT_CAMERA_INFO)" />
    </node>
    <node pkg="nodelet" type="nodelet"
          args="load jsk_pcl/MaskImageFilter manager"
          name="mask_image_filter">
      <remap from="~input" to="$(arg INPUT_POINT_CLOUD)"/>
      <remap from="~input/mask" to="rect_to_mask_image/output"/>
      <remap from="~input/camera_info" to="$(arg INPUT_CAMERA_INFO)" />
    </node>
    <node pkg="nodelet" type="nodelet" name="mask_image_filter_points"
          args="load pcl/ExtractIndices manager">
      <remap from="~input" to="$(arg INPUT_POINT_CLOUD)" />
      <remap from="~indices" to="mask_image_filter/output" />
    </node>
    <!-- normal -->
    <node pkg="nodelet" type="nodelet" name="normal_estimation"
          args="load pcl/NormalEstimationOMP manager">
      <remap from="~input" to="mask_image_filter_points/output"/>
      <rosparam>
        k_search: 0
        radius_search: 0.1
        spatial_locator: 0
      </rosparam>
    </node>

    <node pkg="nodelet" type="nodelet" name="normal_concat"
        args="load jsk_pcl/NormalConcatenater manager">
    <remap from="~input" to="mask_image_filter_points/output" />
    <remap from="~normal" to="normal_estimation/output" />
    <rosparam>
      use_async: false
    </rosparam>
  </node>


    <node pkg="nodelet" type="nodelet"
          args="load jsk_pcl/TorusFinder manager"
          name="handle_detector">
      <remap from="~input" to="normal_concat/output" />
      <rosparam>
        min_radius: 0.17
        max_radius: 0.175
        max_iterations: 1000
        outlier_threshold: 0.01
        use_hint: true
        initial_axis_hint: [0, 0, 1]
        use_normal: true
      </rosparam>
    </node>
    <node pkg="dynamic_tf_publisher" type="tf_publish.py" 
	  name="drive_tf_publisher"
          if="$(arg LAUNCH_DYNAMIC_TF_PUBLISHER)"
	  >
    </node>
    <node pkg="drc_task_common" type="calc_drive_tf.py"
	  name="calc_tf" 
	 >
      <remap from="pose" to="handle_detector/output/pose" />
    </node>
    <node pkg="jsk_interactive_marker" type="transformable_server_sample" 
	  name="handle_server" />
    
  </group>
</launch>
